### PR TITLE
PWN-7169 - debounce amount input

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7169

## Description of Work

debounce amount input

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt [hunk 1]
```diff
@@ -23,4 +23,5 @@ sealed interface SwapStateAction {
 
     object SwitchTokens : SwapStateAction
     object SwapSuccess : SwapStateAction
+    object CancelJobs : SwapStateAction
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/jupiter/statemanager/SwapStateAction.kt:26]: it is not a feature Action name
we really don't have any jobs in swap context, maybe CancelSwap / CancelSwapLoading?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 2]
```diff
@@ -48,13 +52,44 @@ class JupiterSwapPresenter(
...
-        } else {
-            SwapStateAction.TokenAAmountChanged(newAmount)
+        debounceInputJob?.cancel()
+        canselRateJobs()
+        debounceInputJob = launch {
+            val newAmount = amount.toBigDecimalOrZero()
+            val action = if (newAmount.isZero()) {
+                SwapStateAction.EmptyAmountTokenA
+            } else {
+                val pair = featureState?.getTokensPair()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:62]: destructing here doesn't work?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 3]
```diff
@@ -69,7 +82,19 @@ class SwapWidgetMapper {
...
+        tokenAmount: BigDecimal? = null,
+    ): SwapWidgetModel {
+        var result = mapTokenB(token, tokenAmount)
+        if (oldWidgetModel is SwapWidgetModel.Content && oldWidgetModel.fiatAmount != null) {
+            result = result.copy(fiatAmount = oldWidgetModel.fiatAmount)
+        }
+        return result
+    }
+
+    fun mapTokenB(token: SwapTokenModel, tokenAmount: BigDecimal?): SwapWidgetModel.Content {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:97]: maybe we can split logic for TokenB and TokenA mapping into two mappers? Single responsibility :)

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:62]: Снимок экрана 2023-02-24 в 12 12 33
not support nullable

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt (hunk 3 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:97]: mapper doesn't any local props or state, just methods. i don't think its really necessary

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt (hunk 2 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:62]: not support nullable

**Review by kabramovp2p — APPROVED** [review]

**Pull request merged**